### PR TITLE
Add register signup method

### PR DESCRIPTION
**Pull request by racevedoo**
This adds the `registerSignup` method to `IncogniaAPI`

### CODE REVIEW

File: src/main/java/com/incognia/NetworkingClient.java [hunk 1]
```diff
@@ -27,8 +26,7 @@
 
   public NetworkingClient(OkHttpClient httpClient, String baseUrl) {
     this.httpClient = httpClient;
-    this.objectMapper =
-        new ObjectMapper().setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
+    this.objectMapper = ObjectMapperFactory.OBJECT_MAPPER;
```
Comment (racevedoo) [inline src/main/java/com/incognia/NetworkingClient.java:29]: moved to a factory so we can use the same config in our tests

File: src/main/java/com/incognia/SignupResponse.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.Map;
+import java.util.UUID;
+import lombok.Value;
+
+@Value
+public class SignupResponse {
+  UUID id;
+  UUID requestId;
+  Assessment riskAssessment;
+  Map<String, Object> evidence;
```
Comment (racevedoo) [inline src/main/java/com/incognia/SignupResponse.java:12]: decided to keep this as a map at first, for simplicity. we can add getters for some evidences if we need it

File: src/test/java/com/incognia/TokenAwareDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    if ("/api/v2/onboarding/signups".equals(request.getPath())
+        && "POST".equals(request.getMethod())) {
+      assertThat(request.getHeader("Content-Type")).contains("application/json");
+      assertThat(request.getHeader("Authorization")).isEqualTo("Bearer " + token);
+      PostSignupRequestBody postSignupRequestBody =
+          objectMapper.readValue(request.getBody().inputStream(), PostSignupRequestBody.class);
+      assertThat(postSignupRequestBody.getInstallationId()).isEqualTo(expectedInstallationId);
+      assertThat(postSignupRequestBody.getAddressLine()).isEqualTo(expectedAddressLine);
+      String response = getResourceFileAsString("post_onboarding_response.json");
+      return new MockResponse().setResponseCode(200).setBody(response);
```
Comment (joaoaab) [inline src/test/java/com/incognia/TokenAwareDispatcher.java:55]: Super Nit : This Method seems a little bit cluttered for my taste, i would probably add some empty lines in order to better organize the actions in sections of their own, like : 
```suggestion
        && "POST".equals(request.getMethod())) {
      assertThat(request.getHeader("Content-Type")).contains("application/json");
      assertThat(request.getHeader("Authorization")).isEqualTo("Bearer " + token);
      
      PostSignupRequestBody postSignupRequestBody =
          objectMapper.readValue(request.getBody().inputStream(), PostSignupRequestBody.class);
      
      assertThat(postSignupRequestBody.getInstallationId()).isEqualTo(expectedInstallationId);
      assertThat(postSignupRequestBody.getAddressLine()).isEqualTo(expectedAddressLine);
      
      String response = getResourceFileAsString("post_onboarding_response.json");
      return new MockResponse().setResponseCode(200).setBody(response);
```

But i understand this comes completely from taste :smile:

**Review by joaoaab — COMMENTED** [review]
Looks good, i left some small comments about formating and documentation.

File: src/main/java/com/incognia/IncogniaAPI.java [hunk 4]
```diff
@@ -28,14 +25,24 @@ public IncogniaAPI(String clientId, String clientSecret, Region region) {
...
-    this.clientSecret = clientSecret;
-    this.apiUrl = apiUrl;
     // TODO (rato): set client timeout
-    this.httpClient = new OkHttpClient.Builder().build();
+    tokenAwareNetworkingClient =
+        new TokenAwareNetworkingClient(new Builder().build(), apiUrl, clientId, clientSecret);
   }
 
-  // POST onboarding
+  public SignupResponse registerSignup(String installationId, Address address)
```
Comment (figueredo) [inline src/main/java/com/incognia/IncogniaAPI.java:33]: Should we abstract out the idea of response the same way we did for the request? Or the opposite, let the user pass the request object.

Comment (racevedoo) [reply src/main/java/com/incognia/IncogniaAPI.java:33]: we probably shouldn't abstract the response, as the pojos would be equal (or very similar). We can evolve this in the future, tough

**Review by joaoaab — APPROVED** [review]

File: src/test/java/com/incognia/TokenAwareDispatcher.java (hunk 3 above)
Comment (gaabs) [reply src/test/java/com/incognia/TokenAwareDispatcher.java:55]: I believe it would improve its readability if "handle" methods are extracted for each route

File: src/test/java/com/incognia/TokenAwareDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+          .setBody(
+              "{\"access_token\": \""
+                  + token
+                  + "\",\"expires_in\": 100,\"token_type\": \"Bearer\"}");
+    }
+    return new MockResponse().setResponseCode(401);
+  }
+
+  @SneakyThrows
+  static String getResourceFileAsString(String fileName) {
```
Comment (gaabs) [inline src/test/java/com/incognia/TokenAwareDispatcher.java:88]: This could be extracted to a helper/util class

File: src/test/java/com/incognia/IncogniaAPITest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    expectedEvidence.put("location_services", locationServices);
+    expectedEvidence.put("device_integrity", deviceIntegrity);
+
+    assertThat(signupResponse.getEvidence()).containsExactlyInAnyOrderEntriesOf(expectedEvidence);
+  }
+
+  @Test
+  @DisplayName("should throw illegal argument exception with correct message")
+  @SneakyThrows
+  void testRegisterSignup_whenEmptyInstallationId() {
```
Comment (gaabs) [inline src/test/java/com/incognia/IncogniaAPITest.java:80]: You could also add a test checking for null installation id

File: src/main/java/com/incognia/IncogniaAPI.java [hunk 7]
```diff
@@ -28,14 +25,25 @@ public IncogniaAPI(String clientId, String clientSecret, Region region) {
...
+    tokenAwareNetworkingClient =
+        new TokenAwareNetworkingClient(
+            new OkHttpClient.Builder().build(), apiUrl, clientId, clientSecret);
   }
 
-  // POST onboarding
+  public SignupResponse registerSignup(String installationId, Address address)
+      throws IncogniaException {
+    Asserts.assertNotEmpty(installationId, "installation id");
+    Asserts.assertNotNull(address, "address");
```
Comment (gaabs) [inline src/main/java/com/incognia/IncogniaAPI.java:37]: Should we make Address mandatory? Although it is a recommended field, it is currently optional in our docs (and one could easily create an empty Address)

File: src/test/java/com/incognia/TokenAwareDispatcher.java (hunk 3 above)
Comment (joaoaab) [reply src/test/java/com/incognia/TokenAwareDispatcher.java:55]: yeah i believe this would be great too

File: src/main/java/com/incognia/IncogniaAPI.java (hunk 4 above)
Comment (figueredo) [reply src/main/java/com/incognia/IncogniaAPI.java:34]: I meant it regarding naming, not necessarily needing to have a `*Response` being mapped to something else.

I'm a little uncomfortable about having a `*Response` without a `*Request`. As it makes sense to not require the user to build a request object, I was wondering if we could return something that isn't called a response. In the proposal for Ruby, this object is called `SignupAssesment`, what do you think of using the same name?

Comment (figueredo) [reply src/main/java/com/incognia/IncogniaAPI.java:34]: Actually, in Ruby the idea is to return `Incognia::Response` which contains `status` and `resource` (the `SignupAssesment` in this case), but for Java, the status will be implicit in case of success or an exception in case of error, right?

Comment (racevedoo) [reply src/main/java/com/incognia/IncogniaAPI.java:34]: yeah, nice catch. changing to `SignupAssessment`

File: src/main/java/com/incognia/IncogniaAPI.java (hunk 7 above)
Comment (racevedoo) [reply src/main/java/com/incognia/IncogniaAPI.java:37]: It would be quite weird to have a signup without an address. If the user really wants it, they could instantiate an empty address.

File: src/test/java/com/incognia/TokenAwareDispatcher.java (hunk 3 above)
Comment (racevedoo) [reply src/test/java/com/incognia/TokenAwareDispatcher.java]: done!

**Review by gaabs — APPROVED** [review]
